### PR TITLE
Fixes #7844: Integrate ncf generic methods reference into the doc

### PR DESCRIPTION
**Pull request by amousset**
https://www.rudder-project.org/redmine/issues/7844

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -68,6 +71,15 @@ html/$(BASENAME).pdf: $(SOURCES)
 	rm -f *.svg
 	mv $(BASENAME).pdf html/
 
+ncf:
+	git clone https://github.com/Normation/ncf.git
```
Comment (amousset) [inline Makefile]: Same mechanism as rudder-agent repo for man in 3.0+.

**Comment by amousset** [comment]
WIP

**Comment by amousset** [comment]
PR updated

**Comment by amousset** [comment]
Now compatible with python 3, ready for review.

**Comment by amousset** [comment]
PR updated

**Comment by amousset** [comment]
PR updated

File: Makefile [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 
-.PHONY: all clean view
+.PHONY: all clean view ncf-doc
 
 BASENAME = rudder-doc
 SOURCES = $(BASENAME).txt
 TARGETS = epub html pdf readme webhelp webhelp-localsearch
 DOCBOOK_DIST = xsl/xsl-ns-stylesheets
 
+NCF_VERSION = v0.x
```
Comment (jooooooon) [inline Makefile]: Don't forget to do another PR to change this to master in 3.2+

File: tools/ncf_doc_rudder.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  for rudder_version in versions:
+    try:
+      cfengine_version = requests.get(release_info_url + "versions/" + rudder_version + "/components/cfengine").content.decode('ascii')
+    except requests.exceptions.RequestException as e:
+      print(e)
+      sys.exit(1)
+    
+    if cfengine_version in min_version:
+      if StrictVersion(rudder_version) >= StrictVersion(min_version[cfengine_version]):
+        break
```
Comment (jooooooon) [inline tools/ncf_doc_rudder.py]: I don't understand why you have a break statement here. I can see you want to update min_version[cfengine_version] to always have the lowest of rudder_versions that support a given cfengine_version, but this break will *exit* the loop over rudder_versions.

For example, if versions = [ 2.11, 3.0, 3.1, 3.2 ], then you will do this:

  * min_version[3.6] = 2.11
  * break (because 3.0 >= 2.11)

So min_version will only contain 3.6, but not 3.7 (for 3.2).

File: tools/ncf_doc_rudder.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  """Extract minimal Rudder version"""
+  canonified_version = re.findall(r'>=? ?(\d+\.\d+)', version)[0]
+  
+  cfengine_versions = list(min_version.keys())
+  cfengine_versions.sort(key=StrictVersion)
+  
+  for cfengine_version in cfengine_versions:
+    if StrictVersion(canonified_version) <= StrictVersion(cfengine_version):
+      return min_version[cfengine_version]
+  return min_version[cfengine_versions[-1]]
```
Comment (jooooooon) [inline tools/ncf_doc_rudder.py]: I don't get this. It seems to be a default value of "if I don't find any Rudder versions that support CFEngine version N, then return the last Rudder version". But that doesn't make sense - if we have a ncf generic method that only works in CFEngine 3.10, we shouldn't say it works in 3.2 now.

**Comment by jooooooon** [comment]
@amousset Can you please look into these Python scripts questions? It should be quick, then we get this merged and close this topic.

**Comment by amousset** [comment]
PR updated

File: tools/ncf_doc_rudder.py (hunk 4 above)
Comment (amousset) [inline tools/ncf_doc_rudder.py]: Indeed, changed this to not display non-compatible generic methods in the docs.

**Comment by jooooooon** [comment]
Looks great! Thanks!
